pontoon/static/js/widgets/data.js

```diff
@@ -0,0 +1,104 @@
+
+import React from 'react';
+import ReactTable from 'react-table';
+import "react-table/react-table.css";
+
+import {ajax} from 'utils/ajax';
+
+
+export class DataManager extends React.Component {
+
+    constructor (props) {
+        super(props);
+        this.state = {data: []};
+        this.handleResponse = this.handleResponse.bind(this);
+        this.handleSubmit = this.handleSubmit.bind(this);
+    }
+
+    get api () {
+        return this.props.api;
+    }
+
+    get params () {
+        return {};
+    }
+
+    get requestMethod () {
+        return 'get';
+    }
+
+    get submitMethod () {
+        return this.requestMethod;
+    }
+
+    async refreshData () {
+        return this.handleResponse(
+            await ajax[this.requestMethod](this.api, this.params));
+    }
+
+    componentDidMount () {
+        return this.refreshData();
+    }
+
+    async handleResponse (response) {
+        const {status} = response;
+        const json = await response.json();
+        if (status === 200) {
+            const {data} = json;
+            return this.setState({data, errors: {}});
+        }
+        const {errors} = json;
+        return this.setState({errors});
+    }
+
+    async handleSubmit (data) {
+        const params = this.params;
+        params.data = data;
+        return this.handleResponse(
+            await ajax[this.submitMethod](this.api, params));
+    }
+
+    render () {
+        return '';
+    }
+}
+
+
+export class DataTable extends DataManager {
```

Comment:
While this might work, note how inheritance is not idiomatic React and will result foreign to people familiar with it:
https://reactjs.org/docs/composition-vs-inheritance.html

When willing to separate data handling from visuals, the common approach is to externalize state handling in a parent container component, which then passes all the needed data/handlers down to presentational components as props. These presentational components needn't have any state, and hence can be simple functions (provided they don't need lifecycle methods).

There are many articles online explaining all of this, most notably:
* https://medium.com/@dan_abramov/smart-and-dumb-components-7ca2f9a7c7d0
* https://medium.com/@dan_abramov/mixins-are-dead-long-live-higher-order-components-94a0d2f9e750

All in all explicit composition is favored over inheritance.

Reply:
(Another common pattern is to organize components in individual files named after the component, e.g. `DataTable.js`, usually having one component per file.)

Reply:
@julen thanks for review

yep, i am mostly trying to use composition and avoid inheritance for react widgets

in this particular case its what happens in `componentDidMount` that is being inherited more than anything else - so i dont really see how that could be done with composition

for the CheckboxTable that i have also added in this PR it could be done through composition - but still im not entirely convinced of the argument for doing so

Reply:
re organising 1 per file - again im mostly following that - but in some cases i throw them together, expecting that in the future they wiill be separated and then re-exported from index.js

Reply:
> in this particular case its what happens in componentDidMount that is being inherited more than anything else - so i dont really see how that could be done with composition

I believe you are looking for Higher-order Components: https://reactjs.org/docs/higher-order-components.html

Reply:
On that topic of HIC, I've found this article quite interesting: http://americanexpress.io/faccs-are-an-antipattern/